File: serene/src/Serene.Web/Modules/Administration/User/UserForm.cs
namespace Serene.Administration.Forms;

[FormScript("Administration.User")]
[BasedOnRow(typeof(UserRow), CheckNames = true)]
public class UserForm
{
    [LabelWidth(200, UntilNext = true)]
    public string Username { get; set; }
    public string DisplayName { get; set; }
    [EmailAddressEditor]
    public string Email { get; set; }
    [LookupEditor(typeof(RoleRow), Multiple = true)]
    public List<int> Roles { get; set; }
    public string UserImage { get; set; }
    [PasswordEditor, Required(true)]
    public string Password { get; set; }
    [PasswordEditor, Required(true)]
    public string PasswordConfirm { get; set; }
    [OneWay]
    public string Source { get; set; }
    public bool IsActive { get; set;  }
}

File: serene/src/Serene.Web/Modules/ServerTypes/Administration/UserForm.ts
import { BooleanEditor, EmailAddressEditor, ImageUploadEditor, initFormType, LookupEditor, PasswordEditor, PrefixedContext, StringEditor } from "@serenity-is/corelib";

export interface UserForm {
    Username: StringEditor;
    DisplayName: StringEditor;
    Email: EmailAddressEditor;
    Roles: LookupEditor;
    UserImage: ImageUploadEditor;
    Password: PasswordEditor;
    PasswordConfirm: PasswordEditor;
    Source: StringEditor;
    IsActive: BooleanEditor;
}

export class UserForm extends PrefixedContext {
    static readonly formKey = 'Administration.User';
    declare private static init: boolean;

    constructor(...args: ConstructorParameters<typeof PrefixedContext>) {
        super(...args);

        if (!UserForm.init) {
            UserForm.init = true;

            initFormType(UserForm, [
                'Username', StringEditor,
                'DisplayName', StringEditor,
                'Email', EmailAddressEditor,
                'Roles', LookupEditor,
                'UserImage', ImageUploadEditor,
                'Password', PasswordEditor,
                'PasswordConfirm', PasswordEditor,
                'Source', StringEditor,
                'IsActive', BooleanEditor
            ]);
        }
    }
}